File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const { log } = require("console");
let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().split("\n");
const ans = [];
for (let i = 0; i < input.length; i++) {
  const stack = [];
  let text = input[i];
  if (text === ".") break;
  for (let k = 0; k < text.length; k++) {
    if (text[k] === "(" || text[k] === "[") {
      stack.push(text[k]);
    } else if (text[k] === ")") {
      if (stack[stack.length - 1] === "(") {
        stack.pop();
      } else {
        stack.push(text[k]);
      }
    } else if (text[k] === "]") {
      if (stack[stack.length - 1] === "[") {
        stack.pop();
      } else {
        stack.push(text[k]);
      }
    }
  }

  if (stack.length === 0) {
    ans.push("yes");
  } else {
    ans.push("no");
  }
}

console.log(ans.join("\n"));

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        while(true) {
            String str = br.readLine();
            if(str.equals(".")) {
                break;
            }
            Stack<String> stack = new Stack<>();

            String[] sentence = str.split("");
            for (String word : sentence) {
                if (word.equals("(") || word.equals("[")) {
                    stack.push(word);
                } else {
                    if (word.equals(")")) {
                        if (!stack.isEmpty() && stack.peek().equals("(")) {
                            stack.pop();
                        } else {
                            stack.push(word);
                        }
                    }
                    if (word.equals("]")) {
                        if (!stack.isEmpty() && stack.peek().equals("[")) {
                            stack.pop();
                        } else {
                            stack.push(word);
                        }
                    }
                }
            }
            if(stack.isEmpty()) {
                bw.write("yes\n");
            } else {
                bw.write("no\n");
            }
            bw.flush();
        }
    }
}
